starknet/src/utils/stark_eip712.cairo

```diff
@@ -0,0 +1,167 @@
+    proposal_id: u256,
+    choice: Choice,
+    user_voting_strategies: Span<IndexedStrategy>
+) -> felt252 {
+    let mut encoded_data = ArrayTrait::<felt252>::new();
+    VOTE_TYPEHASH.serialize(ref encoded_data);
+    space.serialize(ref encoded_data);
+    voter.serialize(ref encoded_data);
+    proposal_id.struct_hash().serialize(ref encoded_data);
+    choice.serialize(ref encoded_data);
+    user_voting_strategies.struct_hash().serialize(ref encoded_data);
+    hash_typed_data(domain_hash, encoded_data.span().struct_hash(), voter)
+}
+
+fn get_update_proposal_digest(
+    domain_hash: felt252,
+    space: ContractAddress,
+    author: ContractAddress,
+    proposal_id: u256,
+    execution_strategy: @Strategy,
+    salt: felt252
+) -> felt252 {
+    let mut encoded_data = ArrayTrait::<felt252>::new();
+    UPDATE_PROPOSAL_TYPEHASH.serialize(ref encoded_data);
+    space.serialize(ref encoded_data);
+    author.serialize(ref encoded_data);
+    proposal_id.struct_hash().serialize(ref encoded_data);
+    execution_strategy.struct_hash().serialize(ref encoded_data);
+    salt.serialize(ref encoded_data);
+    hash_typed_data(domain_hash, encoded_data.span().struct_hash(), author)
+}
+
+fn get_domain_hash(name: felt252, version: felt252) -> felt252 {
+    let mut encoded_data = ArrayTrait::<felt252>::new();
+    DOMAIN_TYPEHASH.serialize(ref encoded_data);
+    name.serialize(ref encoded_data);
+    version.serialize(ref encoded_data);
+    get_tx_info().unbox().chain_id.serialize(ref encoded_data);
+    get_contract_address().serialize(ref encoded_data);
+    encoded_data.span().struct_hash()
+}
+
+fn hash_typed_data(
+    domain_hash: felt252, message_hash: felt252, signer: ContractAddress
+) -> felt252 {
+    let mut encoded_data = ArrayTrait::<felt252>::new();
+    STARKNET_MESSAGE.serialize(ref encoded_data);
+    domain_hash.serialize(ref encoded_data);
+    signer.serialize(ref encoded_data);
+    message_hash.serialize(ref encoded_data);
+    encoded_data.span().struct_hash()
+}
+
+/// Verifies the signature of a message by calling the account contract.
+fn verify_signature(
+    digest: felt252, signature: Array<felt252>, account: ContractAddress, account_type: felt252
+) {
+    if account_type == 'snake' {
+        assert(
+            AccountCamelABIDispatcher {
+                contract_address: account
+            }.supportsInterface(ERC165_ACCOUNT_INTERFACE_ID) == true,
+            'Invalid Account'
+        );
+        AccountCamelABIDispatcher { contract_address: account }.isValidSignature(digest, signature);
+    } else if account_type == 'camel' {
+        assert(
+            AccountABIDispatcher {
+                contract_address: account
+            }.supports_interface(ERC165_OLD_ACCOUNT_INTERFACE_ID) == true,
+            'Invalid Account'
+        );
+        AccountABIDispatcher { contract_address: account }.is_valid_signature(digest, signature);
+    } else {
+        panic_with_felt252('Invalid Account Type');
```

Comment:
excessive complexity maybe?

Reply:
I actually think it'd be better because then it would be enforced when the tx is created / cast.. like it would instantly fail instead of going through all the code and failing here!